### PR TITLE
Extend use of BatchMetricsLogger to all autologging integrations

### PR DESCRIPTION
**Pull request by mohamad-arabi**
Signed-off-by: Mohamad Arabi <mohamad.arabi@databricks.com>

## What changes are proposed in this pull request?

Integrate the use of `BatchMetricsLogger ` for all model flavors, such that metrics are batch logged throughout training phase.

## How is this patch tested?

Added test cases for each relevant ML flavor.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tensorflow.py [hunk 1]
```diff
@@ -651,10 +656,6 @@ class __MLflowTfKerasCallback(Callback):
         Records model structural information as params after training finishes.
         """
 
-        def __init__(self, log_models):
-            super().__init__()
```
Comment (mohamad-arabi) [inline mlflow/tensorflow.py:655]: Note: Since `log_models` is not added to `self`  in `__init__` anymore, removing the entire method so that super's init is invoked automatically without running into `Useless super delegation in method` linting error.

File: mlflow/tensorflow.py [hunk 2]
```diff
@@ -719,10 +720,6 @@ class __MLflowTfKeras2Callback(Callback):
         Records model structural information as params when training starts.
         """
 
-        def __init__(self, log_models):
```
Comment (mohamad-arabi) [inline mlflow/tensorflow.py:722]: Note: Since `log_models` is not added to `self`  in `__init__` anymore, removing the entire method so that super's init is invoked automatically without running into `Useless super delegation in method` linting error.

File: tests/xgboost/test_xgboost_autolog.py [hunk 3]
```diff
@@ -238,6 +274,53 @@ def test_xgb_autolog_logs_metrics_with_early_stopping(bst_params, dtrain):
             assert metric_history == evals_result[eval_name][metric_name] + [best_metrics]
 
 
+@pytest.mark.large
+def test_xgb_autolog_batch_metrics_logger_logs_metrics_with_early_stopping(bst_params, dtrain):
```
Comment (dbczumar) [inline tests/xgboost/test_xgboost_autolog.py:278]: During some of our offline discussion, I think we suggested `spy`ing on the `BatchMetricsLogger` so that metrics are logged to the run as well. That way, we can compare `patched_metrics_data` to `run.data.metrics` and ensure that the same keys are present. This, coupled with 1-2 concrete value checks (e.g., `"best_iteration" in metric_dict`), should be sufficient.

Can we update our test cases to follow this paradigm? For setting up spies, the `wraps` keyword argument on `patch()` should be helpful: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock (`wraps` is also defined on `patch()` but only seems to be documented here)

File: tests/xgboost/test_xgboost_autolog.py [hunk 4]
```diff
@@ -238,6 +274,53 @@ def test_xgb_autolog_logs_metrics_with_early_stopping(bst_params, dtrain):
...
+    for eval_name in [e[1] for e in evals]:
+        for metric_name in params["eval_metric"]:
+            metric_key = "{}-{}".format(eval_name, metric_name)
+            metric_history = [x[1] for x in patched_metrics_data if x[0] == metric_key]
+
+            assert metric_key in metric_dict
+            assert len(metric_history) == 20 + 1
+
+            best_metrics = evals_result[eval_name][metric_name][model.best_iteration]
+            assert metric_history == evals_result[eval_name][metric_name] + [best_metrics]
```
Comment (dbczumar) [inline tests/xgboost/test_xgboost_autolog.py:321]: On second thought, I don't think we need to check the metric history this rigorously. Presumably, we do this in other test cases in the suite. Our goal here is just to make sure that `batch_metrics_logger` is being used; we can be reasonably confident that this is the case by comparing metric names in `patched_metrics_data`  entries with `run.data.metrics`.

**Review by dbczumar — CHANGES_REQUESTED** [review]
@mohamad-arabi Implementation looks great - left a couple comments around testing!

File: tests/xgboost/test_xgboost_autolog.py (hunk 3 above)
Comment (dbczumar) [reply tests/xgboost/test_xgboost_autolog.py:278]: EDIT: My suggestion to use `patch()` is actually misguided. Instead, we should update the mocked method to call the original method. We can do this by grabbing a reference to the original method before we apply the patch.

File: mlflow/pytorch/_pytorch_autolog.py [hunk 5]
```diff
@@ -49,152 +49,151 @@ def _autolog(log_every_n_epoch=1, log_models=True):
...
+            def on_test_end(self, trainer, pl_module):
+                """
+                Logs accuracy and other relevant metrics on the testing end
+
+                :param trainer: pytorch lightning trainer instance
+                :param pl_module: pytorch lightning base module
+                """
+                try_mlflow_log(mlflow.set_tag, "Mode", "testing")
+                for key, value in trainer.callback_metrics.items():
+                    try_mlflow_log(mlflow.log_metric, key, float(value))
```
Comment (dbczumar) [inline mlflow/pytorch/_pytorch_autolog.py:145]: Should we use `record_metrics` here? (Probably not too significant since this is presumably called at the end of training)

File: mlflow/fastai.py [hunk 6]
```diff
@@ -295,63 +300,64 @@ def autolog(log_models=True):
...
+                        f.write(summary)
+                    try_mlflow_log(mlflow.log_artifact, local_path=summary_file)
+                finally:
+                    shutil.rmtree(tempdir)
+
+            def on_train_end(self, **kwargs):
+                if log_models:
+                    try_mlflow_log(log_model, self.learner, artifact_path="model")
+
+        return __MLflowFastaiCallback
```
Comment (dbczumar) [inline mlflow/fastai.py:360]: Nit: Can we have these `get*Callback` methods across the various modules construct an instance of the callback class, or is there a case where we need to return the class definition instead?

File: tests/fastai/test_fastai_autolog.py [hunk 7]
```diff
@@ -262,3 +264,34 @@ def test_fastai_autolog_non_early_stop_callback_does_not_log(fastai_random_data_
...
+            patched_metrics_data.extend(metrics.items())
+            original(self, metrics, step)
+
+        record_metrics_mock.side_effect = record_metrics_side_effect
+        _, run = fastai_random_data_run_with_callback(
+            iris_data(), fit_variant, manual_run, callback, patience
+        )
+
+    patched_metrics_data = dict(patched_metrics_data)
+    original_metrics = run.data.metrics
```
Comment (dbczumar) [inline tests/fastai/test_fastai_autolog.py:294]: Can we add a small check to each of these test cases across the various modules?:

- Pick *one* concrete metric name (e.g., `train_loss`) and assert that the metric name is in `run.data.metrics`.

This assertion will guard against the case that our mocking logic somehow resulted in metrics being swallowed or inadvertently mutated. Looks like we do the this with Keras and TF Keras for  `restored_epoch` and `stopped_epoch`, which is awesome; let's add it to the other ones!

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py [hunk 8]
```diff
@@ -285,6 +287,47 @@ def test_tf_keras_autolog_early_stop_logs(tf_keras_random_data_run_with_callback
     assert history.history["loss"][restored_epoch] == metric_history[-1].value
 
 
+@pytest.mark.large
+@pytest.mark.parametrize("restore_weights", [True])
+@pytest.mark.parametrize("callback", ["early"])
+@pytest.mark.parametrize("patience", [0, 1, 5])
+def test_tf_keras_autolog_batch_metrics_logger_logs_expected_metrics(
```
Comment (dbczumar) [inline tests/tensorflow_autolog/test_tensorflow2_autolog.py:294]: Is there a non-tf-keras test case that we should add? Vanilla TensorFlow workflows are also important here.

**Review by dbczumar — COMMENTED** [review]
@mohamad-arabi This is very close! Left a few small comments.

File: mlflow/pytorch/_pytorch_autolog.py (hunk 5 above)
Comment (mohamad-arabi) [reply mlflow/pytorch/_pytorch_autolog.py:145]: No, because after extensive  reading of pytorch docs, from my understanding: testing stage which invokes `on_test_end ` is not part of the training stage whether it be `fit` or `train`. Therefore the batch metrics logger context would have been torn down, and recording metrics using it would not make sense. I'm sure you would know better, if I can guarantee that `on_test_end()` will be called before the context is teared down then definitely we should use `record_metrics()` instead of `log_metric `.

File: mlflow/fastai.py (hunk 6 above)
Comment (mohamad-arabi) [reply mlflow/fastai.py:360]: Yes we can, will update in commit

Comment (mohamad-arabi) [reply mlflow/fastai.py:360]: Only pytorch we need to return class, because we check `isinstance(callbacks, __MLflowPLCallback)`.

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py (hunk 8 above)
Comment (mohamad-arabi) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py:315]: No. We only add batch metrics logger to tf.keras `fit` and `fit_generator`.  We don't use the logger for Vanilla TensorFlow. For `tf.estimator` we use TensorBoard to log metrics, and  that already has a custom batching solution implemented `_metric_queue`. Correct me if I am wrong

Comment (dbczumar) [reply tests/tensorflow_autolog/test_tensorflow2_autolog.py:315]: Right! We had this discussion previously. Thanks for the reminder!

File: mlflow/pytorch/_pytorch_autolog.py [hunk 9]
```diff
@@ -49,152 +49,157 @@ def _autolog(log_every_n_epoch=1, log_models=True):
...
+
+            def on_train_end(self, trainer, pl_module):
+                """
+                Logs the model checkpoint into mlflow - models folder on the training end
+
+                :param trainer: pytorch lightning trainer instance
+                :param pl_module: pytorch lightning base module
+                """
+                # manually flushing any remaining metrics from training.
+                metrics_logger._flush()
```
Comment (dbczumar) [inline mlflow/pytorch/_pytorch_autolog.py:130]: Can we make this a public API on `BatchMetricsLogger` (i.e. drop the `_`) since we have legitimate use cases for calling it outside of the module where it's defined?

Comment (mohamad-arabi) [reply mlflow/pytorch/_pytorch_autolog.py]: @dbczumar  Now that I made it a public API do you think I need to add a unti test for `flush()` in `test_autologging_utils.py`?

File: mlflow/utils/autologging_utils.py [hunk 10]
```diff
@@ -156,7 +156,16 @@ def resolve_input_example_and_signature(
...
 
 class BatchMetricsLogger:
-    def __init__(self, run_id):
+    """
+    The BatchMetricsLogger will log metrics in batch against an mlflow run.
+    If run_id is passed to to constructor then all recording and logging will
+    happen against that run_id.
+    If no run_id is passed into constructor, then it is assumed that there is
+    an active mlflow run when there is a need to log_batch metrics. The id of that active run
+    will be retrieved from mlflow.
```
Comment (dbczumar) [inline mlflow/utils/autologging_utils.py:165]:
```suggestion
    If no run_id is passed into constructor, then the run ID will be fetched
    from `mlflow.active_run()` each time `record_metrics()` is called; in this
    case, callers must ensure that an active run is present before invoking
    `record_metrics()`.
```

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/3723/files/1655615b27d4e4721555b102bfc7133c62f9e67d..bcbda87059117e443d1880caace0709977ce87ac#r534889780 is addressed & tests pass!
